exec/cnex/lib/string.c

```diff
@@ -0,0 +1,142 @@
+#include "string.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "array.h"
+#include "cell.h"
+#include "exec.h"
+#include "global.h"
+#include "nstring.h"
+#include "number.h"
+#include "stack.h"
+
+// ToDo: UTF8 string operations on all functions.
+
+
+void string_find(TExecutor *exec)
+{
+    TString *t = peek(exec->stack, 0)->string;
+    TString *s = peek(exec->stack, 1)->string;
+
+    int64_t ret = string_findString(s, 0, t);
+
+    pop(exec->stack);
+    pop(exec->stack);
+    push(exec->stack, cell_fromNumber(number_from_sint64(ret)));
+}
+
+void string_lower(TExecutor *exec)
+{
+    TString *ss = top(exec->stack)->string;
+
+    Cell *r = cell_newCellType(cString);
+    // Note: string_toLowerCase() creates a new string; it does not modify the existing string in-place.
+    // ToDo: Would it be better to modify the string in place?  This should be tested and benchmarked.
+    r->string = string_toLowerCase(ss);
+
+    pop(exec->stack);
+    push(exec->stack, r);
+}
+
+void string_split(TExecutor *exec)
+{
+    TString *d = peek(exec->stack, 0)->string; // Delimeter char or string
+    TString *s = peek(exec->stack, 1)->string; // Original String to split
+
+    Cell *r = cell_createArrayCell(0);
+    size_t i = 0;
+    while (i < s->length) {
+        Cell e = { 0, .type = cString };
+        int64_t nd = string_findString(s, i, d);
+        if (nd == -1) {
+            e.string = string_subString(s, i, s->length - i);
+            cell_arrayAppendElement(r, e);
+            string_freeString(e.string);
+            break;
+        } else if ((size_t)nd > i) {
+            e.string = string_subString(s, i, nd-i);
+            cell_arrayAppendElement(r, e);
+            string_freeString(e.string);
+        }
+        i = nd + d->length;
+    }
+    pop(exec->stack);
+    pop(exec->stack);
+
+    push(exec->stack, r);
+}
+
+void string_splitLines(TExecutor *exec)
+{
+    TString *s = peek(exec->stack, 0)->string; // Original String to split
+
+    TString *eol = string_createCString("\r\n");
```

Comment:
You don't need to allocate/free a new one of these. Create a local `struct TString`, point the pointer at some constant data, set the length, and then you don't need to free it.

Reply:
Fixed.  Probably a lot of other places I should be doing this type of stuff in, too.